MFVNeutralino/plugins/Vertexer.cc

```diff
@@ -1,891 +1,1769 @@
+        const size_t vtxidx0 = sv_ascending_vtxidx[i];
+        reco::Vertex& sv0 = vertices->at(vtxidx0);
+        double sv0x = sv0.x() - bsx;
+        double sv0y = sv0.y() - bsy;
+        double phi0 = atan2(sv0y, sv0x);
+        if (verbose){
+          printf("-----loop # %lu ----- \n", i);
+          printf(" sv0'idx: %lu \n", vtxidx0);
+          printf(" sv0'ntrack: %u \n", sv0.nTracks());
+        }
+        for (size_t j = 0; j < sv_ascending_vtxidx.size(); j++) {
+          if (i == j) continue;
+          const size_t vtxidx1 = sv_ascending_vtxidx[j];
+          reco::Vertex& sv1 = vertices->at(vtxidx1);
+          if (verbose){
+            printf(" sv1'idx: %lu \n", vtxidx1);
+            printf(" sv1'ntrack: %u \n", sv1.nTracks());
+          }
+
+          // only consider a pair with at least 3 tracks per vertex
+          if (sv0.nTracks() > 2 && sv1.nTracks() > 2) {
+
+            std::pair<bool, std::vector<std::vector<size_t>>> sharedjet_tool = sharedjets(vtxidx0, vtxidx1, sv_match_tracktojet_which_jetidx, sv_match_tracktojet_which_trkidx);
+
+            // loop thru {1,1}+{1,n} nsharedjets and remove just one shared track from v0 if a |dPhi(v0,one shared track)| > pi/2
+            if (sharedjet_tool.first) {
+              if (verbose)
+                printf("start shj implementation to {1,1} and {1,n} \n");
+              std::vector<std::vector<size_t>> sv_lonesharedtrack_which_trkidx = sharedjet_tool.second;
+              std::vector<size_t> sv0_lonesharedtrack_which_trkidx = sv_lonesharedtrack_which_trkidx[0];
+              std::vector<size_t> sv1_lonesharedtrack_which_trkidx = sv_lonesharedtrack_which_trkidx[1];
+              if (verbose) {
+                printf("size of set of lone shared tracks per sv0: %lu \n", sv0_lonesharedtrack_which_trkidx.size());
+                printf("size of set of lone shared tracks per sv1: %lu \n", sv1_lonesharedtrack_which_trkidx.size());
+              }
+              for (size_t k = 0; k < sv0_lonesharedtrack_which_trkidx.size(); k++) {
+                track_vec tks_sv0 = sv_total_track_which_trk_vec[vtxidx0];
+                size_t idx = sv0_lonesharedtrack_which_trkidx[k];
+                h_resolve_shared_jets_lonetrkvtx_dphi->Fill(fabs(reco::deltaPhi(tks_sv0[idx]->phi(), phi0)));
+
+                // drop the lone track pointing backwards from the vertex direction!
+                if (fabs(reco::deltaPhi(tks_sv0[idx]->phi(), phi0)) > M_PI / 2) {
+                  eraseElement(sv_total_track_which_trkidx[vtxidx0], idx);
+                }
+              }
+              if (verbose) {
+                printf("sv0'idx: %lu with ntrack before: %u",vtxidx0, sv0.nTracks());
+                printf("sv1'idx: %lu with ntrack before: %u",vtxidx1, sv1.nTracks());
+              }
+
+              track_set  sv0_resolved_sharedtracks_trkset;
+              for (unsigned int i = 0; i < sv_total_track_which_trkidx[vtxidx0].size(); ++i) {

```

Comment:
This looks like a potential bug to me. `i` is already defined up [here](https://github.com/DisplacedVertices/cmssw-usercode/pull/5/files?w=1#diff-eaef89c993d668ab99b433c8e0844dbd1b7a8260a673b2d83468451a634fa1caR1423). It would be good to give a different name here, and to double check throughout the code that no other variable names are reused within the loop. It might be good to use unique names for the index variable names in these loops to avoid this potential issue.

Reply:
To follow up: I turned on a different compiler option than we usually use, and am copying the warnings potentially relevant to this for Vertexer.cc. Some may not be related to the new implementation, of course, but take a look!

```
/uscms/homes/j/joeyr/PR/mfv_946p1/src/JMTucker/MFVNeutralino/plugins/Vertexer.cc: In member function 'virtual void MFVVertexer::produce(edm::Event&, const edm::EventSetup&)':
/uscms/homes/j/joeyr/PR/mfv_946p1/src/JMTucker/MFVNeutralino/plugins/Vertexer.cc:878:21: warning: declaration of 'v' shadows a previous local [-Wshadow]
       reco::Vertex& v = vertices->at(i);^M
                     ^
/uscms/homes/j/joeyr/PR/mfv_946p1/src/JMTucker/MFVNeutralino/plugins/Vertexer.cc:589:39: note: shadowed declaration is here
   std::vector<reco::Vertex>::iterator v[2];^M
                                       ^
/uscms/homes/j/joeyr/PR/mfv_946p1/src/JMTucker/MFVNeutralino/plugins/Vertexer.cc:889:24: warning: declaration of 'r' shadows a previous local [-Wshadow]
       for (const auto& r : vertex_track_set(v)) {^M
                        ^
/uscms/homes/j/joeyr/PR/mfv_946p1/src/JMTucker/MFVNeutralino/plugins/Vertexer.cc:888:20: note: shadowed declaration is here
       const double r = mag(vx, vy, vz);^M
                    ^
/uscms/homes/j/joeyr/PR/mfv_946p1/src/JMTucker/MFVNeutralino/plugins/Vertexer.cc:1312:22: warning: declaration of 'v' shadows a previous local [-Wshadow]
         reco::Vertex v = potential_merged_vertices[i];^M
                      ^
/uscms/homes/j/joeyr/PR/mfv_946p1/src/JMTucker/MFVNeutralino/plugins/Vertexer.cc:589:39: note: shadowed declaration is here
   std::vector<reco::Vertex>::iterator v[2];^M
                                       ^
/uscms/homes/j/joeyr/PR/mfv_946p1/src/JMTucker/MFVNeutralino/plugins/Vertexer.cc:1476:33: warning: declaration of 'i' shadows a previous local [-Wshadow]
               for (unsigned int i = 0; i < sv_total_track_which_trkidx[vtxidx0].size(); ++i) {^M
                                 ^
/uscms/homes/j/joeyr/PR/mfv_946p1/src/JMTucker/MFVNeutralino/plugins/Vertexer.cc:1424:19: note: shadowed declaration is here
       for (size_t i = 0; i < sv_ascending_vtxidx.size(); i++) {^M
                   ^
/uscms/homes/j/joeyr/PR/mfv_946p1/src/JMTucker/MFVNeutralino/plugins/Vertexer.cc: In member function 'std::pair<bool, std::vector<std::vector<long unsigned int> > > MFVVertexer::sharedjets(int, int, const std::vector<std::vector<int> >&, const std::vector<std::vector<int> >&)':
/uscms/homes/j/joeyr/PR/mfv_946p1/src/JMTucker/MFVNeutralino/plugins/Vertexer.cc:1580:17: warning: declaration of 'it' shadows a previous local [-Wshadow]
       for (auto it = sv0_m.begin(); it != sv0_m.end(); ) {^M
                 ^~
/uscms/homes/j/joeyr/PR/mfv_946p1/src/JMTucker/MFVNeutralino/plugins/Vertexer.cc:1568:34: note: shadowed declaration is here
       std::vector<int>::iterator it = getFirstCommonElement(sv_match_tracktojet_which_temp_jetidx[vtx0idx], sv_match_tracktojet_which_temp_jetidx[vtx1idx]);^M
                                  ^~
/uscms/homes/j/joeyr/PR/mfv_946p1/src/JMTucker/MFVNeutralino/plugins/Vertexer.cc:1599:17: warning: declaration of 'it' shadows a previous local [-Wshadow]
       for (auto it = sv1_m.begin(); it != sv1_m.end(); ) {^M
                 ^~
/uscms/homes/j/joeyr/PR/mfv_946p1/src/JMTucker/MFVNeutralino/plugins/Vertexer.cc:1568:34: note: shadowed declaration is here
       std::vector<int>::iterator it = getFirstCommonElement(sv_match_tracktojet_which_temp_jetidx[vtx0idx], sv_match_tracktojet_which_temp_jetidx[vtx1idx]);^M
                                  ^~
/uscms/homes/j/joeyr/PR/mfv_946p1/src/JMTucker/MFVNeutralino/plugins/Vertexer.cc: In member function 'void MFVVertexer::fillCommonOutputHists(std::unique_ptr<std::vector<reco::Vertex> >&, const reco::Vertex&, edm::ESHandle<TransientTrackBuilder>&, size_t)':
/uscms/homes/j/joeyr/PR/mfv_946p1/src/JMTucker/MFVNeutralino/plugins/Vertexer.cc:1713:22: warning: declaration of 'r' shadows a previous local [-Wshadow]
     for (const auto& r : vertex_track_set(v)) {^M
                      ^
/uscms/homes/j/joeyr/PR/mfv_946p1/src/JMTucker/MFVNeutralino/plugins/Vertexer.cc:1712:18: note: shadowed declaration is here
     const double r = mag(vx, vy, vz);^M
                  ^
```

Reply:
This should be taken care of but you can double-check.

Reply:
Done.